# Toronto Neighborhood Clustering Analysis

In [1]:
import numpy as np 

import pandas as pd 
pd.set_option('display.max_columns', None)
pd.set_option('display.max_rows', None)

from bs4 import BeautifulSoup

import json # library to handle JSON files

#!conda install -c conda-forge geopy --yes # uncomment this line if you haven't completed the Foursquare API lab
from geopy.geocoders import Nominatim # convert an address into latitude and longitude values

import requests # library to handle requests
from pandas.io.json import json_normalize # tranform JSON file into a pandas dataframe

# Matplotlib and associated plotting modules
import matplotlib.cm as cm
import matplotlib.colors as colors

# import k-means from clustering stage
from sklearn.cluster import KMeans

!conda install -c conda-forge folium=0.5.0 --yes # uncomment this line if you haven't completed the Foursquare API lab
import folium # map rendering library

print('Libraries imported.')

Solving environment: done

## Package Plan ##

  environment location: /opt/conda/envs/Python36

  added / updated specs: 
    - folium=0.5.0


The following packages will be downloaded:

    package                    |            build
    ---------------------------|-----------------
    ca-certificates-2019.6.16  |       hecc5488_0         145 KB  conda-forge
    branca-0.3.1               |             py_0          25 KB  conda-forge
    altair-3.1.0               |           py36_0         724 KB  conda-forge
    folium-0.5.0               |             py_0          45 KB  conda-forge
    vincent-0.4.4              |             py_1          28 KB  conda-forge
    certifi-2019.6.16          |           py36_1         149 KB  conda-forge
    openssl-1.1.1c             |       h516909a_0         2.1 MB  conda-forge
    ------------------------------------------------------------
                                           Total:         3.2 MB

The following NEW packages will be 

# Part 1 - Wikipedia Scrape and Dataframe Build

Using BeautifulSoup we get the [table of Toronto neighborhods by area code](https://en.wikipedia.org/wiki/List_of_postal_codes_of_Canada:_M) from Wikipedia. We will build the pandas Data Frame 'df" which countains the Postcode, Borough, and Neighborhoods from this table.

In [3]:
website_url = requests.get('https://en.wikipedia.org/wiki/List_of_postal_codes_of_Canada:_M').text
soup = BeautifulSoup(website_url,'lxml')
table = soup.find('table',{'class':'wikitable sortable'})
ths = table.find_all('th')

headings = [th.text.strip() for th in ths]
headings[2] = 'Neighborhood'

Table_Data = []
for tr in table.find_all('tr'):
    tds = tr.find_all('td')
    if not tds:
        continue
    Postcode, Borough, Neighborhood = [td.text.strip() for td in tds[:3]]
    Table_Data.append([Postcode, Borough, Neighborhood])
    
Table_Data_Grouped = []
for row in Table_Data:
    entry = []

df = pd.DataFrame(Table_Data)
df.columns = headings 

df = df[df.Borough != 'Not assigned'] # This removes all entries which have no Borough name

df.loc[df['Neighborhood'] == 'Not assigned','Neighborhood'] = df['Borough'] # This replaces Neighborhood name with Borough name if Neighborhood name is not assigned

neib = lambda a: ", ".join(a)
df = df.groupby(by='Postcode').agg({'Borough': 'max','Neighborhood': neib}).reset_index() # This groups neighborhoods with the same postcode/borough

df.head()

Postcode      Borough                            Neighborhood
0      M1B  Scarborough                          Rouge, Malvern
1      M1C  Scarborough  Highland Creek, Rouge Hill, Port Union
2      M1E  Scarborough       Guildwood, Morningside, West Hill
3      M1G  Scarborough                                  Woburn
4      M1H  Scarborough                               Cedarbrae

In [4]:
df.shape

(103, 3)

# Part 2 - Geospatial data and Foursquare API

Using [geographical data](https://cocl.us/Geospatial_data) we will append two additional columns for 'Latitude' and 'Longitude' of each borough to the dataframe, and rename the dataframe 'df2.'

In [5]:
url = 'https://cocl.us/Geospatial_data'
geo=pd.read_csv(url)
df2 = pd.concat([df, geo['Latitude'], geo['Longitude']], axis=1, sort=False)
df2.head()

Postcode      Borough                            Neighborhood   Latitude  \
0      M1B  Scarborough                          Rouge, Malvern  43.806686   
1      M1C  Scarborough  Highland Creek, Rouge Hill, Port Union  43.784535   
2      M1E  Scarborough       Guildwood, Morningside, West Hill  43.763573   
3      M1G  Scarborough                                  Woburn  43.770992   
4      M1H  Scarborough                               Cedarbrae  43.773136   

   Longitude  
0 -79.194353  
1 -79.160497  
2 -79.188711  
3 -79.216917  
4 -79.239476

Using folium, we will construct a map of Toronto.  We will superimpose neighborhood markers on top of the map using data from 'df2.' Neighborhoods will be grouped by postcode. 

In [8]:
address = 'Toronto, ON'

geolocator = Nominatim(user_agent="Toronto_explorer")
location = geolocator.geocode(address)
latitude = location.latitude
longitude = location.longitude
# create map of Toronto using latitude and longitude values
map_toronto = folium.Map(location=[latitude, longitude], zoom_start=10)

# add markers to map
for lat, lng, borough in zip(df2['Latitude'], df2['Longitude'], df2['Neighborhood']):
    label = '{}'.format(borough)
    label = folium.Popup(label, parse_html=True)
    folium.CircleMarker(
        [lat, lng],
        radius=5,
        popup=label,
        color='blue',
        fill=True,
        fill_color='#3186cc',
        fill_opacity=0.7,
        parse_html=False).add_to(map_toronto)  
    
map_toronto

We will query the Foursquare API for venue data corresponding to the Toronto boroughs in 'df2.'  The function below runs through the boroughs in 'df2' and queries the API, appending the query results to venues_list.  These will be aggregated into the dataframe toronto_venues.

In [12]:
CLIENT_ID = 'WQZ4C3RX1UX3YLTANCQFQ4XMOD0W5LSMRHJSTCGXJQ0VEX1L' # your Foursquare ID
CLIENT_SECRET = 'BX10JIX4Q0R3JACDMIQXFHINLCOXPUDWNWERJOCVFCGZRWX5' # your Foursquare Secret
VERSION = '20190719' # Foursquare API version

def getNearbyVenues(names, latitudes, longitudes, radius=500):
    
    venues_list=[]
    for name, lat, lng in zip(names, latitudes, longitudes):
            
        # create the API request URL
        url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}&limit={}'.format(
            CLIENT_ID, 
            CLIENT_SECRET, 
            VERSION, 
            lat, 
            lng, 
            radius, 
            LIMIT)
            
        # make the GET request
        results = requests.get(url).json()["response"]['groups'][0]['items']
        
        # return only relevant information for each nearby venue
        venues_list.append([(
            name, 
            lat, 
            lng, 
            v['venue']['name'], 
            v['venue']['location']['lat'], 
            v['venue']['location']['lng'],  
            v['venue']['categories'][0]['name']) for v in results])

    nearby_venues = pd.DataFrame([item for venue_list in venues_list for item in venue_list])
    nearby_venues.columns = ['Neighborhood', 
                  'Latitude', 
                  'Longitude', 
                  'Venue', 
                  'Venue Latitude', 
                  'Venue Longitude', 
                  'Venue Category']
    
    return(nearby_venues)

LIMIT = 100

radius = 500

toronto_venues = getNearbyVenues(names=df2['Neighborhood'],
                                   latitudes=df2['Latitude'],
                                   longitudes=df2['Longitude']
                                  )


toronto_venues.head()

Neighborhood   Latitude  Longitude  \
0                          Rouge, Malvern  43.806686 -79.194353   
1                          Rouge, Malvern  43.806686 -79.194353   
2  Highland Creek, Rouge Hill, Port Union  43.784535 -79.160497   
3       Guildwood, Morningside, West Hill  43.763573 -79.188711   
4       Guildwood, Morningside, West Hill  43.763573 -79.188711   

                             Venue  Venue Latitude  Venue Longitude  \
0                          Wendy's       43.807448       -79.199056   
1            Interprovincial Group       43.805630       -79.200378   
2            Royal Canadian Legion       43.782533       -79.163085   
3  Swiss Chalet Rotisserie & Grill       43.767697       -79.189914   
4                G & G Electronics       43.765309       -79.191537   

         Venue Category  
0  Fast Food Restaurant  
1            Print Shop  
2                   Bar  
3           Pizza Place  
4     Electronics Store

In [10]:
print('There are {} uniques categories.'.format(len(toronto_venues['Venue Category'].unique())))

There are 269 uniques categories.


The individula neighborhoods with now be one-hot encoded with respect to the 269 unique venue categries, and a new dataframe will be created with the venue information.  We will name this dataframe 'toronto_onehot.'  Then we will group by neighborhood into the dataframe 'toronto_grouped.'

In [69]:
# one hot encoding
toronto_onehot = pd.get_dummies(toronto_venues[['Venue Category']], prefix="", prefix_sep="")

# add neighborhood column back to dataframe
toronto_onehot['Neighborhood'] = toronto_venues['Neighborhood'] 

# move neighborhood column to the first column
fixed_columns = [toronto_onehot.columns[-1]] + list(toronto_onehot.columns[:-1])
toronto_onehot = toronto_onehot[fixed_columns]

toronto_grouped = toronto_onehot.groupby('Neighborhood').mean().reset_index()
toronto_grouped.head()


Neighborhood  Yoga Studio  \
0                           Adelaide, King, Richmond          0.0   
1                                          Agincourt          0.0   
2  Agincourt North, L'Amoreaux East, Milliken, St...          0.0   
3  Albion Gardens, Beaumond Heights, Humbergate, ...          0.0   
4                             Alderwood, Long Branch          0.0   

   Accessories Store  Afghan Restaurant  Airport  Airport Food Court  \
0                0.0                0.0      0.0                 0.0   
1                0.0                0.0      0.0                 0.0   
2                0.0                0.0      0.0                 0.0   
3                0.0                0.0      0.0                 0.0   
4                0.0                0.0      0.0                 0.0   

   Airport Gate  Airport Lounge  Airport Service  Airport Terminal  \
0           0.0             0.0              0.0               0.0   
1           0.0             0.0              0.0               0.0   
2           0.0             0.0              0.0               0.0   
3           0.0             0.0              0.0               0.0   
4           0.0             0.0              0.0               0.0   

   American Restaurant  Antique Shop  Aquarium  Art Gallery  Art Museum  \
0                 0.03           0.0       0.0         0.01        0.01   
1                 0.00           0.0       0.0         0.00        0.00   
2                 0.00           0.0       0.0         0.00        0.00   
3                 0.00           0.0       0.0         0.00        0.00   
4                 0.00           0.0       0.0         0.00        0.00   

   Arts & Crafts Store  Asian Restaurant  Athletics & Sports  Auto Garage  \
0                  0.0              0.03                 0.0          0.0   
1                  0.0              0.00                 0.0          0.0   
2                  0.0              0.00                 0.0          0.0   
3                  0.0              0.00                 0.0          0.0   
4                  0.0              0.00                 0.1          0.0   

   Auto Workshop  BBQ Joint  Baby Store  Bagel Shop  Bakery  Bank   Bar  \
0            0.0        0.0         0.0         0.0    0.02   0.0  0.04   
1            0.0        0.0         0.0         0.0    0.00   0.0  0.00   
2            0.0        0.0         0.0         0.0    0.00   0.0  0.00   
3            0.0        0.0         0.0         0.0    0.00   0.0  0.00   
4            0.0        0.0         0.0         0.0    0.00   0.0  0.00   

   Baseball Field  Baseball Stadium  Basketball Court  Basketball Stadium  \
0             0.0               0.0               0.0                 0.0   
1             0.0               0.0               0.0                 0.0   
2             0.0               0.0               0.0                 0.0   
3             0.0               0.0               0.0                 0.0   
4             0.0               0.0               0.0                 0.0   

   Beach  Beer Bar  Beer Store  Belgian Restaurant  Bike Shop  Bistro  \
0    0.0       0.0    0.000000                 0.0        0.0     0.0   
1    0.0       0.0    0.000000                 0.0        0.0     0.0   
2    0.0       0.0    0.000000                 0.0        0.0     0.0   
3    0.0       0.0    0.111111                 0.0        0.0     0.0   
4    0.0       0.0    0.000000                 0.0        0.0     0.0   

   Board Shop  Boat or Ferry  Bookstore  Boutique  Brazilian Restaurant  \
0         0.0            0.0       0.01       0.0                  0.01   
1         0.0            0.0       0.00       0.0                  0.00   
2         0.0            0.0       0.00       0.0                  0.00   
3         0.0            0.0       0.00       0.0                  0.00   
4         0.0            0.0       0.00       0.0                  0.00   

   Breakfast Spot  Brewery  Bridal Shop  Bubble Tea S

For illustration purposes, we will show the top five venue categories in each borough below. 

In [70]:
num_top_venues = 5

for hood in toronto_grouped['Neighborhood']:
    print("----"+hood+"----")
    temp = toronto_grouped[toronto_grouped['Neighborhood'] == hood].T.reset_index()
    temp.columns = ['venue','freq']
    temp = temp.iloc[1:]
    temp['freq'] = temp['freq'].astype(float)
    temp = temp.round({'freq': 2})
    print(temp.sort_values('freq', ascending=False).reset_index(drop=True).head(num_top_venues))
    print('\n')

----Adelaide, King, Richmond----
             venue  freq
0      Coffee Shop  0.07
1             Café  0.05
2              Bar  0.04
3  Thai Restaurant  0.04
4       Steakhouse  0.04


----Agincourt----
            venue  freq
0  Clothing Store  0.25
1  Breakfast Spot  0.25
2          Lounge  0.25
3  Sandwich Place  0.25
4   Metro Station  0.00


----Agincourt North, L'Amoreaux East, Milliken, Steeles East----
                       venue  freq
0                 Playground   0.5
1                       Park   0.5
2  Middle Eastern Restaurant   0.0
3                      Motel   0.0
4        Monument / Landmark   0.0


----Albion Gardens, Beaumond Heights, Humbergate, Jamestown, Mount Olive, Silverstone, South Steeles, Thistletown----
                  venue  freq
0         Grocery Store  0.22
1              Pharmacy  0.11
2           Pizza Place  0.11
3  Fast Food Restaurant  0.11
4            Beer Store  0.11


----Alderwood, Long Branch----
                venue  freq
0         Pizza

In [71]:
def return_most_common_venues(row, num_top_venues):
    row_categories = row.iloc[1:]
    row_categories_sorted = row_categories.sort_values(ascending=False)
    
    return row_categories_sorted.index.values[0:num_top_venues]

In [73]:
num_top_venues = 10

indicators = ['st', 'nd', 'rd']

# create columns according to number of top venues
columns = ['Neighborhood']
for ind in np.arange(num_top_venues):
    try:
        columns.append('{}{} Most Common Venue'.format(ind+1, indicators[ind]))
    except:
        columns.append('{}th Most Common Venue'.format(ind+1))

# create a new dataframe
boroughs_venues_sorted = pd.DataFrame(columns=columns)
boroughs_venues_sorted['Neighborhood'] = toronto_grouped['Neighborhood']

for ind in np.arange(toronto_grouped.shape[0]):
    boroughs_venues_sorted.iloc[ind, 1:] = return_most_common_venues(toronto_grouped.iloc[ind, :], num_top_venues)

boroughs_venues_sorted    

Neighborhood  \
0                            Adelaide, King, Richmond   
1                                           Agincourt   
2   Agincourt North, L'Amoreaux East, Milliken, St...   
3   Albion Gardens, Beaumond Heights, Humbergate, ...   
4                              Alderwood, Long Branch   
5     Bathurst Manor, Downsview North, Wilson Heights   
6                                     Bayview Village   
7                   Bedford Park, Lawrence Manor East   
8                                         Berczy Park   
9                         Birch Cliff, Cliffside West   
10  Bloordale Gardens, Eringate, Markland Wood, Ol...   
11       Brockton, Exhibition Place, Parkdale Village   
12  Business Reply Mail Processing Centre 969 Eastern   
13                        CFB Toronto, Downsview East   
14  CN Tower, Bathurst Quay, Island airport, Harbo...   
15                        Cabbagetown, St. James Town   
16                                Caledonia-Fairbanks   
17              Canada Post Gateway Processing Centre   
18                                          Cedarbrae   
19                                 Central Bay Street   
20          Chinatown, Grange Park, Kensington Market   
21                                           Christie   
22                               Church and Wellesley   
23                    Clairlea, Golden Mile, Oakridge   
24            Clarks Corners, Sullivan, Tam O'Shanter   
25    Cliffcrest, Cliffside, Scarborough Village West   
26  Cloverdale, Islington, Martin Grove, Princess ...   
27                     Commerce Court, Victoria Hotel   
28                                         Davisville   
29                                   Davisville North   
30  Deer Park, Forest Hill SE, Rathnelly, South Hi...   
31     Del Ray, Keelesdale, Mount Dennis, Silverthorn   
32           Design Exchange, Toronto Dominion Centre   
33                                    Don Mills North   
34  Dorset Park, Scarborough Town Centre, Wexford ...   
35                       Dovercourt Village, Dufferin   
36                                  Downsview Central   
37                                Downsview Northwest   
38                                     Downsview West   
39                 Downsview, North Park, Upwood Park   
40        East Birchmount Park, Ionview, Kennedy Park   
41                                       East Toronto   
42                                   Emery, Humberlea   
43                       Fairview, Henry Farm, Oriole   
44             First Canadian Place, Underground city   
45                   Flemingdon Park, Don Mills South   
46                Forest Hill North, Forest Hill West   
47                                          Glencairn   
48                  Guildwood, Morningside, West Hill   
49                     Harbord, University of Toronto   
50  Harbourfront East, Toronto Islands, Union Station   
51                          Harbourfront, Regent Park   
52                      High Park, The Junction South   
53             Highland Creek, Rouge Hill, Port Union   
54                                  Hillcrest Village   
55       Humber Bay Shores, Mimico South, New Toronto   
56  Humber Bay, King's Mill Park, Kingsway Park So...   
57                                      Humber Summit   
58                                 Humewood-Cedarvale   
59  Kingsview Village, Martin Grove Gardens, Richv...   
60  Kingsway Park South West, Mimico NW, The Queen...   
61                                    L'Amoreaux West   
62                   Lawrence Heights, Lawrence Manor   
63                                      Lawrence Park   
64                                            Leaside   
65                           Little Portugal, Trinity   
66                                  Maryvale, Wexford   
67                        Moore Park, Summerhill East   
68                                 North Toronto West   
69                                          Northwe

# Part 3 - Clustering and Segmentation of Neighborhoods

Now we will segement and cluster the neighborhood into 5 clusters determined by the venues appearing commonly. This will be done using k-means clustering.

In [74]:
# set number of clusters
kclusters = 4

toronto_grouped_clustering = toronto_grouped.drop('Neighborhood', 1)

# run k-means clustering
kmeans = KMeans(n_clusters=kclusters, random_state=0).fit(toronto_grouped_clustering)

# check cluster labels generated for each row in the dataframe
kmeans.labels_[0:102] 

array([0, 0, 2, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 2, 0, 0, 2, 0, 0, 0, 0, 0,
       0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 2, 2, 0, 2, 3, 0,
       0, 0, 2, 0, 0, 0, 0, 0, 0, 1, 0, 0, 3, 0, 0, 2, 0, 0, 0, 2, 0, 0,
       0, 2, 0, 0, 0, 0, 2, 0, 2, 0, 0, 0, 0, 2, 0, 0, 0, 0, 0, 0, 0, 0,
       2, 0, 0, 0, 0, 0, 0, 0, 0, 0, 2], dtype=int32)

In [75]:
toronto_grouped_clustering.head()

Yoga Studio  Accessories Store  Afghan Restaurant  Airport  \
0          0.0                0.0                0.0      0.0   
1          0.0                0.0                0.0      0.0   
2          0.0                0.0                0.0      0.0   
3          0.0                0.0                0.0      0.0   
4          0.0                0.0                0.0      0.0   

   Airport Food Court  Airport Gate  Airport Lounge  Airport Service  \
0                 0.0           0.0             0.0              0.0   
1                 0.0           0.0             0.0              0.0   
2                 0.0           0.0             0.0              0.0   
3                 0.0           0.0             0.0              0.0   
4                 0.0           0.0             0.0              0.0   

   Airport Terminal  American Restaurant  Antique Shop  Aquarium  Art Gallery  \
0               0.0                 0.03           0.0       0.0         0.01   
1               0.0                 0.00           0.0       0.0         0.00   
2               0.0                 0.00           0.0       0.0         0.00   
3               0.0                 0.00           0.0       0.0         0.00   
4               0.0                 0.00           0.0       0.0         0.00   

   Art Museum  Arts & Crafts Store  Asian Restaurant  Athletics & Sports  \
0        0.01                  0.0              0.03                 0.0   
1        0.00                  0.0              0.00                 0.0   
2        0.00                  0.0              0.00                 0.0   
3        0.00                  0.0              0.00                 0.0   
4        0.00                  0.0              0.00                 0.1   

   Auto Garage  Auto Workshop  BBQ Joint  Baby Store  Bagel Shop  Bakery  \
0          0.0            0.0        0.0         0.0         0.0    0.02   
1          0.0            0.0        0.0         0.0         0.0    0.00   
2          0.0            0.0        0.0         0.0         0.0    0.00   
3          0.0            0.0        0.0         0.0         0.0    0.00   
4          0.0            0.0        0.0         0.0         0.0    0.00   

   Bank   Bar  Baseball Field  Baseball Stadium  Basketball Court  \
0   0.0  0.04             0.0               0.0               0.0   
1   0.0  0.00             0.0               0.0               0.0   
2   0.0  0.00             0.0               0.0               0.0   
3   0.0  0.00             0.0               0.0               0.0   
4   0.0  0.00             0.0               0.0               0.0   

   Basketball Stadium  Beach  Beer Bar  Beer Store  Belgian Restaurant  \
0                 0.0    0.0       0.0    0.000000                 0.0   
1                 0.0    0.0       0.0    0.000000                 0.0   
2                 0.0    0.0       0.0    0.000000                 0.0   
3                 0.0    0.0       0.0    0.111111                 0.0   
4                 0.0    0.0       0.0    0.000000                 0.0   

   Bike Shop  Bistro  Board Shop  Boat or Ferry  Bookstore  Boutique  \
0        0.0     0.0         0.0            0.0       0.01       0.0   
1        0.0     0.0         0.0            0.0       0.00       0.0   
2        0.0     0.0         0.0            0.0       0.00       0.0   
3        0.0     0.0         0.0            0.0       0.00       0.0   
4        0.0     0.0         0.0            0.0       0.00       0.0   

   Brazilian Restaurant  Breakfast Spot  Brewery  Bridal Shop  \
0                  0.01            0.02      0.0          0.0   
1                  0.00            0.25      0.0          0.0   
2                  0.00            0.00      0.0          0.0   
3                  0.00            0.00      0.0          0.0   
4                  0.00            0.00      0.0          0.0   

   Bubble Tea Shop  Building  Burger Joint  Burrito Place  Bus Line  \
0              0.0     

In [100]:
# add clustering labels
boroughs_venues_sorted.drop('Cluster_Labels',1,inplace = True)
boroughs_venues_sorted.insert(0,'Cluster_Labels', kmeans.labels_)

toronto_merged = df2

# merge toronto_grouped with toronto_data to add latitude/longitude for each neighborhood
toronto_merged = toronto_merged.join(boroughs_venues_sorted.set_index('Neighborhood'), on='Neighborhood')


#Remove any rows that contain NaN 
toronto_merged.dropna(inplace = True)
toronto_merged.reset_index(drop = True, inplace = True)

# Change cluster labels from floats to int
toronto_merged.Cluster_Labels = toronto_merged.Cluster_Labels.astype(int)

toronto_merged

Postcode           Borough  \
0       M1B       Scarborough   
1       M1C       Scarborough   
2       M1E       Scarborough   
3       M1G       Scarborough   
4       M1H       Scarborough   
5       M1J       Scarborough   
6       M1K       Scarborough   
7       M1L       Scarborough   
8       M1M       Scarborough   
9       M1N       Scarborough   
10      M1P       Scarborough   
11      M1R       Scarborough   
12      M1S       Scarborough   
13      M1T       Scarborough   
14      M1V       Scarborough   
15      M1W       Scarborough   
16      M2H        North York   
17      M2J        North York   
18      M2K        North York   
19      M2N        North York   
20      M2P        North York   
21      M2R        North York   
22      M3A        North York   
23      M3B        North York   
24      M3C        North York   
25      M3H        North York   
26      M3J        North York   
27      M3K        North York   
28      M3L        North York   
29      M3M        North York   
30      M3N        North York   
31      M4A        North York   
32      M4B         East York   
33      M4C         East York   
34      M4E      East Toronto   
35      M4G         East York   
36      M4H         East York   
37      M4J         East York   
38      M4K      East Toronto   
39      M4L      East Toronto   
40      M4M      East Toronto   
41      M4N   Central Toronto   
42      M4P   Central Toronto   
43      M4R   Central Toronto   
44      M4S   Central Toronto   
45      M4T   Central Toronto   
46      M4V   Central Toronto   
47      M4W  Downtown Toronto   
48      M4X  Downtown Toronto   
49      M4Y  Downtown Toronto   
50      M5A  Downtown Toronto   
51      M5B  Downtown Toronto   
52      M5C  Downtown Toronto   
53      M5E  Downtown Toronto   
54      M5G  Downtown Toronto   
55      M5H  Downtown Toronto   
56      M5J  Downtown Toronto   
57      M5K  Downtown Toronto   
58      M5L  Downtown Toronto   
59      M5M        North York   
60      M5N   Central Toronto   
61      M5P   Central Toronto   
62      M5R   Central Toronto   
63      M5S  Downtown Toronto   
64      M5T  Downtown Toronto   
65      M5V  Downtown Toronto   
66      M5W  Downtown Toronto   
67      M5X  Downtown Toronto   
68      M6A        North York   
69      M6B        North York   
70      M6C              York   
71      M6E              York   
72      M6G  Downtown Toronto   
73      M6H      West Toronto   
74      M6J      West Toronto   
75      M6K      West Toronto   
76      M6L        North York   
77      M6M              York   
78      M6N              York   
79      M6P      West Toronto   
80      M6R      West Toronto   
81      M6S      West Toronto   
82      M7A      Queen's Park   
83      M7R       Mississauga   
84      M7Y      East Toronto   
85      M8V         Etobicoke   
86      M8W         Etobicoke   
87      M8X         Etobicoke   
88      M8Y         Etobicoke   
89      M8Z         Etobicoke   
90      M9B         Etobicoke   
91      M9C         Etobicoke   
92      M9L        North York   
93      M9M        North York   
94      M9N              York   
95      M9P         Etobicoke   
96      M9R         Etobicoke   
97      M9V         Etobicoke   
98      M9W         Etobicoke   

                                         Neighborhood   Latitude  Longitude  \
0                                      Rouge, Malvern  43.806686 -79.194353   
1              Highland Creek, Rouge Hill, Port Union  43.784535 -79.160497   
2                   Guildwood, Morningside, West Hill  43.763573 -79.188711   
3                                              Woburn  43.770992 -79.216917   
4                                           Cedarbrae  43.773136 -79.239476   
5                                 Scarborough Village  43.744734 -79.239476   
6         East Birchmount Park, Ionview, Kennedy Park  43.727929 -79.262029   
7                     Clairlea, Golden Mile, Oakridge  43.711112 -79.2

To visualize this clustering, we will redraw the map of Toronto and color the borough markers according to cluster. 

In [102]:
# create map
map_clusters = folium.Map(location=[latitude, longitude], zoom_start=11)

# set color scheme for the clusters
x = np.arange(kclusters)
ys = [i + x + (i*x)**2 for i in range(kclusters)]
colors_array = cm.rainbow(np.linspace(0, 1, len(ys)))
rainbow = [colors.rgb2hex(i) for i in colors_array]

# add markers to the map
markers_colors = []
for lat, lon, poi, cluster in zip(toronto_merged['Latitude'], toronto_merged['Longitude'], toronto_merged['Neighborhood'], toronto_merged['Cluster_Labels']):
    label = folium.Popup(str(poi) + ' Cluster ' + str(cluster), parse_html=True)
    folium.CircleMarker(
        [lat, lon],
        radius=5,
        popup=label,
        color=rainbow[cluster-1],
        fill=True,
        fill_color=rainbow[cluster-1],
        fill_opacity=0.7).add_to(map_clusters)
       
map_clusters

We can confirm the cluster information below. 

# Cluster 0

In [23]:
toronto_merged.loc[toronto_merged['Cluster Labels'] == 0, toronto_merged.columns[[1] + list(range(5, toronto_merged.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue  \
0        Scarborough               0    Chinese Restaurant   
1        Scarborough               0    Chinese Restaurant   
2        Scarborough               0    Chinese Restaurant   
3        Scarborough               0    Chinese Restaurant   
4        Scarborough               0    Chinese Restaurant   
5        Scarborough               0    Chinese Restaurant   
6        Scarborough               0    Chinese Restaurant   
7        Scarborough               0    Chinese Restaurant   
8        Scarborough               0    Chinese Restaurant   
9        Scarborough               0    Chinese Restaurant   
10       Scarborough               0    Chinese Restaurant   
11       Scarborough               0    Chinese Restaurant   
12       Scarborough               0    Chinese Restaurant   
13       Scarborough               0    Chinese Restaurant   
14       Scarborough               0    Chinese Restaurant   
15       Scarborough               0    Chinese Restaurant   
16       Scarborough               0    Chinese Restaurant   
17        North York               0           Coffee Shop   
18        North York               0           Coffee Shop   
19        North York               0           Coffee Shop   
20        North York               0           Coffee Shop   
21        North York               0           Coffee Shop   
22        North York               0           Coffee Shop   
23        North York               0           Coffee Shop   
24        North York               0           Coffee Shop   
25        North York               0           Coffee Shop   
26        North York               0           Coffee Shop   
27        North York               0           Coffee Shop   
28        North York               0           Coffee Shop   
29        North York               0           Coffee Shop   
30        North York               0           Coffee Shop   
31        North York               0           Coffee Shop   
32        North York               0           Coffee Shop   
33        North York               0           Coffee Shop   
34        North York               0           Coffee Shop   
37      East Toronto               0      Greek Restaurant   
41      East Toronto               0      Greek Restaurant   
42      East Toronto               0      Greek Restaurant   
43      East Toronto               0      Greek Restaurant   
50  Downtown Toronto               0           Coffee Shop   
51  Downtown Toronto               0           Coffee Shop   
52  Downtown Toronto               0           Coffee Shop   
53  Downtown Toronto               0           Coffee Shop   
54  Downtown Toronto               0           Coffee Shop   
55  Downtown Toronto               0           Coffee Shop   
56  Downtown Toronto               0           Coffee Shop   
57  Downtown Toronto               0           Coffee Shop   
58  Downtown Toronto               0           Coffee Shop   
59  Downtown Toronto               0           Coffee Shop   
60  Downtown Toronto               0           Coffee Shop   
61  Downtown Toronto               0           Coffee Shop   
62        North York               0           Coffee Shop   
66  Downtown Toronto               0           Coffee Shop   
67  Downtown Toronto               0           Coffee Shop   
68  Downtown Toronto               0           Coffee Shop   
69  Downtown Toronto               0           Coffee Shop   
70  Downtown Toronto               0           Coffee Shop   
71        North York               0           Coffee Shop   
72        North York               0           Coffee Shop   
75  Downtown Toronto               0           Coffee Shop   
76      West Toronto               0                   Bar   
77      West Toronto               0                   Bar   
78      West Toronto               0                   Bar   
79        North York               0         

# Cluster 1

In [25]:
toronto_merged.loc[toronto_merged['Cluster Labels'] == 1, toronto_merged.columns[[1] + list(range(5, toronto_merged.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue  \
35         East York               1           Coffee Shop   
36         East York               1           Coffee Shop   
38         East York               1           Coffee Shop   
39         East York               1           Coffee Shop   
40         East York               1           Coffee Shop   
44   Central Toronto               1           Coffee Shop   
45   Central Toronto               1           Coffee Shop   
46   Central Toronto               1           Coffee Shop   
47   Central Toronto               1           Coffee Shop   
48   Central Toronto               1           Coffee Shop   
49   Central Toronto               1           Coffee Shop   
63   Central Toronto               1           Coffee Shop   
64   Central Toronto               1           Coffee Shop   
65   Central Toronto               1           Coffee Shop   
88         Etobicoke               1           Pizza Place   
89         Etobicoke               1           Pizza Place   
90         Etobicoke               1           Pizza Place   
91         Etobicoke               1           Pizza Place   
92         Etobicoke               1           Pizza Place   
93         Etobicoke               1           Pizza Place   
94         Etobicoke               1           Pizza Place   
95         Etobicoke               1           Pizza Place   
99         Etobicoke               1           Pizza Place   
100        Etobicoke               1           Pizza Place   
101        Etobicoke               1           Pizza Place   
102        Etobicoke               1           Pizza Place   

    2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
35            Pizza Place                  Park              Pharmacy   
36            Pizza Place                  Park              Pharmacy   
38            Pizza Place                  Park              Pharmacy   
39            Pizza Place                  Park              Pharmacy   
40            Pizza Place                  Park              Pharmacy   
44            Pizza Place        Sandwich Place                  Park   
45            Pizza Place        Sandwich Place                  Park   
46            Pizza Place        Sandwich Place                  Park   
47            Pizza Place        Sandwich Place                  Park   
48            Pizza Place        Sandwich Place                  Park   
49            Pizza Place        Sandwich Place                  Park   
63            Pizza Place        Sandwich Place                  Park   
64            Pizza Place        Sandwich Place                  Park   
65            Pizza Place        Sandwich Place                  Park   
88         Sandwich Place           Coffee Shop              Pharmacy   
89         Sandwich Place           Coffee Shop              Pharmacy   
90         Sandwich Place           Coffee Shop              Pharmacy   
91         Sandwich Place           Coffee Shop              Pharmacy   
92         Sandwich Place           Coffee Shop              Pharmacy   
93         Sandwich Place           Coffee Shop              Pharmacy   
94         Sandwich Place           Coffee Shop              Pharmacy   
95         Sandwich Place           Coffee Shop              Pharmacy   
99         Sandwich Place           Coffee Shop              Pharmacy   
100        Sandwich Place           Coffee Shop              Pharmacy   
101        Sandwich Place           Coffee Shop              Pharmacy   
102        Sandwich Place           Coffee Shop              Pharmacy   

    5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
35                   Bank          Burger Joint   Sporting Goods Shop   
36                   Bank          Burger Joint   Sporting Goods Shop   
38                   Bank          Burger Joint   Sporting Goods Shop   
39                   Bank          Burger Joint   Sporting Goods Shop   
4

# Cluster 2

In [26]:
toronto_merged.loc[toronto_merged['Cluster Labels'] == 2, toronto_merged.columns[[1] + list(range(5, toronto_merged.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue 2nd Most Common Venue  \
73    York               2                  Park     Convenience Store   
74    York               2                  Park     Convenience Store   
80    York               2                  Park     Convenience Store   
81    York               2                  Park     Convenience Store   
98    York               2                  Park     Convenience Store   

   3rd Most Common Venue 4th Most Common Venue 5th Most Common Venue  \
73               Brewery  Fast Food Restaurant                 Field   
74               Brewery  Fast Food Restaurant                 Field   
80               Brewery  Fast Food Restaurant                 Field   
81               Brewery  Fast Food Restaurant                 Field   
98               Brewery  Fast Food Restaurant                 Field   

   6th Most Common Venue 7th Most Common Venue 8th Most Common Venue  \
73        Sandwich Place        Discount Store            Restaurant   
74        Sandwich Place        Discount Store            Restaurant   
80        Sandwich Place        Discount Store            Restaurant   
81        Sandwich Place        Discount Store            Restaurant   
98        Sandwich Place        Discount Store            Restaurant   

   9th Most Common Venue 10th Most Common Venue  
73  Caribbean Restaurant            Coffee Shop  
74  Caribbean Restaurant            Coffee Shop  
80  Caribbean Restaurant            Coffee Shop  
81  Caribbean Restaurant            Coffee Shop  
98  Caribbean Restaurant            Coffee Shop

# Cluster 3

In [ ]:
toronto_merged.loc[toronto_merged['Cluster Labels'] == 3, toronto_merged.columns[[1] + list(range(5, toronto_merged.shape[1]))]]

# Cluster 4

In [27]:
toronto_merged.loc[toronto_merged['Cluster Labels'] == 4, toronto_merged.columns[[1] + list(range(5, toronto_merged.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue 2nd Most Common Venue  \
85  Queen's Park               4           Coffee Shop                  Park   

   3rd Most Common Venue 4th Most Common Venue 5th Most Common Venue  \
85                   Gym           Yoga Studio         Smoothie Shop   

   6th Most Common Venue 7th Most Common Venue 8th Most Common Venue  \
85    Seafood Restaurant        Sandwich Place          Burger Joint   

   9th Most Common Venue 10th Most Common Venue  
85         Burrito Place                   Café